PyDSS/registry.py

```diff
@@ -0,0 +1,188 @@
+"""Manages PyDSS customized modules."""
+
+from collections import defaultdict
+import logging
+import os
+import pathlib
+
+import PyDSS
+from PyDSS.common import ControllerType, CONTROLLER_TYPES
+from PyDSS.exceptions import InvalidParameter
+from PyDSS.utils.utils import dump_data, load_data
+
+
+DEFAULT_REGISTRY = {
+    "Controllers": {
+        ControllerType.PV_CONTROLLER.value: [
+            {
+                "name": "cpf",
+                "filename": os.path.join(
+                    os.path.dirname(getattr(PyDSS, "__path__")[0]),
+                    "PyDSS/pyControllers/Controllers/Settings/PvControllers.toml"
+                ),
+            },
+            {
+                "name": "volt-var",
+                "filename": os.path.join(
+                    os.path.dirname(getattr(PyDSS, "__path__")[0]),
+                    "PyDSS/pyControllers/Controllers/Settings/PvControllers.toml"
+                ),
+            },
+        ],
+        ControllerType.SOCKET_CONTROLLER.value: [],
+        ControllerType.STORAGE_CONTROLLER.value: [],
+        ControllerType.XMFR_CONTROLLER.value: [],
+        ControllerType.MOTOR_STALL.value: [],
+        ControllerType.PV_VOLTAGE_RIDETHROUGH.value: [],
+        ControllerType.FAULT_CONTROLLER.value: [],
+    },
+}
+
+REQUIRED_CONTROLLER_FIELDS = ("name", "filename")
+
+
+logger = logging.getLogger(__name__)
+
+
+class Registry:
+    """Manages controllers registered with PyDSS."""
+    _REGISTRY_FILENAME = ".pydss-registry.json"
+
+    def __init__(self, registry_filename=None):
+        if registry_filename is None:
+            self._registry_filename = os.path.join(
+                str(pathlib.Path.home()),
+                self._REGISTRY_FILENAME,
+            )
+        else:
+            self._registry_filename = registry_filename
+
+        self._controllers = {x: {} for x in CONTROLLER_TYPES}
+        if not os.path.exists(self._registry_filename):
+            self.reset_defaults()
+        else:
+            data = load_data(self._registry_filename)
+            for controller_type in data["Controllers"]:
+                for controller in data["Controllers"][controller_type]:
+                    self._add_controller(controller_type, controller)
+
+    def _add_controller(self, controller_type, controller):
+        name = controller["name"]
+        filename = controller["filename"]
+        if self.is_controller_registered(controller_type, name):
+            raise InvalidParameter(f"{controller_type} / {name} is already registered")
```

Comment:
Instead of raise an error, would the registration be skipped if it's already there, and show user a message?

Reply:
The names would be the same but they could point to different controller setting files.  I could check the contents and throw an error if they are different.  I think I prefer the simple approach of throwing an error, for now.

Reply:
It makes sense, I agree.